Show survivor count and vote events on infoscreen

tiittapauppi opened the pull request.

tiittapauppi committed:
In progress

tiittapauppi committed:
Infoscreen Survivor count template added

tiittapauppi committed:
FleetSocket added

tiittapauppi committed:
Moved Survivor count from News to Info

tiittapauppi committed:
Open votes to infoboard

nicou commented on src/routes/infoboard.js:
```diff
@@ -94,6 +94,7 @@ router.get('/', handleAsyncErrors(async (req, res) => {
  * @returns {InfoEntry.model} 200 - Inserted InfoEntry values
  */
 router.put('/', handleAsyncErrors(async (req, res) => {
+	console.log(req.body + " " + res)
```
Leftover console.log

nicou commented on src/rules/social/infoboard.js:
```diff
@@ -0,0 +1,34 @@
+
+async function updateInfoentryScheduledToClose() {
+	const activeInfoentry = await new InfoEntry().where('enabled', true).where('active_until', '!=', '').fetchAll();
+	console.log("### Löytyi tuloksia")
```
Leftover console.log

nicou commented on src/rules/social/infoboard.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+async function updateInfoentryScheduledToClose() {
+	const activeInfoentry = await new InfoEntry().where('enabled', true).where('active_until', '!=', '').fetchAll();
```
Probably should check for non-null values instead of non-empty strings:
```suggestion
	const activeInfoentry = await new InfoEntry().where('enabled', true).whereNotNull('active_until').fetchAll();
```

tiittapauppi replied:
Tested, doesn't work. Crashes the backend.

```
        const activeInfoentry = await new InfoEntry().where('enabled', true).whe
reNotNull('active_until').fetchAll();
                                                                      ^
TypeError: (intermediate value).where(...).whereNotNull is not a function
    at updateInfoentryScheduledToClose (/workspace/src/rules/social/infoboard.js
:15:71)
```

nicou replied:
Yeah, was thinking of Knex syntax 😅 Probably should be `.where('active_until', '!=', null)`, but haven't tested it.

japsu replied:
In Postgres, `null is null` is `true` but `null = null` is `null`, which is falsy. However, also `null != null` and `null <> null` are `null`.

Usually one would use `x is null`/`x is not null` to check for nullity. If you need to treat `null` as a comparable value, the correct comparison operator is the monstrosity known as `is not distinct from`. https://www.postgresql.org/docs/current/functions-comparison.html

Haven't used this particular ORM but if you actually need a nullity comparison, my best guess would be `.where('active_until', 'is not', null)`.

That being said, if `active_until` has a default value that is the empty string `''`, then `.where('active_until', '!=', '')` is indeed the correct solution.

nicou commented on src/rules/social/infoboard.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+// Update infoentrys that are scheduled to close
+interval(updateInfoentryScheduledToClose, 10000);
```
```suggestion
interval(updateInfoentryScheduledToClose, POLL_FREQUENCY_MS);
```

nicou commented on src/store/storeSocket.ts:
```diff
@@ -2,13 +2,18 @@ import { throttle } from 'lodash';
 function sendDataChanges(io: SocketIoServer | SocketIoNamespace) {
 	const currentData = store.getState().data;
 
+	io.to(`/fleet`).emit('fleetUpdate');
```
Is this needed? 🤔

tiittapauppi replied:
Yes, is needed. Without this survivor count won't update on infoboard when there are changes in the fleet data and survivor count. Tested.

tiittapauppi committed:
Update infoboard.js

Removed log

tiittapauppi reviewed (commented):
Checked all comments, this is now OK for me.

tiittapauppi committed:
Voting open and voting results to infoboard

tiittapauppi committed:
Updating query for closing infoentries

nicou committed:
PR improvements

nicou committed:
Simplify display of total souls alive

Update the infoboard entry with the correct survivors count every 10
seconds so that no frontend changes are needed.

nicou committed:
Only show enabled infoboard entries

nicou committed:
Create new vote infoboard entry in backend
